cli: add filtering support for --list_* options

JosephDavisC opened the pull request:
### Title
cli: add filtering support for --list_probes and --list_detectors (fix #1362)

### Description
This PR adds **filtering support** to the `--list_probes` and `--list_detectors` CLI options.
- Users can now filter by **plugin family** (e.g., `dan`) or by a **specific plugin** (e.g., `dan.AntiDAN`).
- If no matches are found, the CLI will display a clear message (e.g., *“No detectors match the provided 'detector_spec': misleading.Invalid”*) and exit with a non-zero code.

This resolves issue #1362.

### Verification
Steps to verify:
-  Run tests: `pytest -q tests/test_cli_list_filtering.py` (all pass).
-  Check filtering works for probes:
  ```bash
  garak --list_probes -p dan
  garak --list_probes -p dan,dan.AntiDAN
  ```
-  Check filtering works for detectors:
  ```bash
  garak --list_detectors -d misleading
  garak --list_detectors -d misleading.Invalid
  ```
-  Confirm invalid spec exits with error message.

JosephDavisC commented:
I have read the DCO Document and I hereby sign the DCO

JosephDavisC commented:
recheck

leondz reviewed (changes requested):
Thanks for this. Would prefer to retain `print_plugins` functionality for probes & detectors. Can we factor this filtering up into that function, perhaps also covering other plugin types?

JosephDavisC commented:
Hey there, just a heads-up, I accidentally clicked re-request review 😅. Still working on the changes, I’ll push updates soon and ping again when it’s ready!

JosephDavisC commented:
Just finished making all the requested fixes. Hope everything is good now! Let me know if there’s anything else I should tweak. Thanks for the feedback!

jmartin-tech commented on garak/command.py:
```diff
@@ -168,14 +203,30 @@ def print_plugins(prefix: str, color):
 
 def print_probes():
     from colorama import Fore
+    from garak import _config
+
+    filtered_plugins = None
+
+    if hasattr(_config.transient, "cli_args") and _config.transient.cli_args:
+        probe_spec = getattr(_config.transient.cli_args, "probes", None)
+        if probe_spec and probe_spec.lower() not in ("", "auto", "all", "*"):
+            filtered_plugins, _ = _config.parse_plugin_spec(probe_spec, "probes")
```
`command` should not need to process `_config`, this work is already accomplished in `cli.py`:

https://github.com/NVIDIA/garak/blob/736dd134147404252041dad95a0059c16487eee3/garak/cli.py#L553-L574

The `parsed_specs` from that code block are what the issue is requesting be printed in context.  The noted section of the configuration parsing could move up in `cli.py` to before call to `print_probes()` and detection of whether the combination of arguments should result in printing the unfiltered list or `parsed_specs["probe"]` can be evaluated there.

This could looks something like:
``` Python
        elif args.list_probes:
            selected_probes = None
            if args.probes or _cli_config_supplied:
                selected_probes=parsed_specs["probe"]
            command.print_probes(selected=selected_probes)
```

`cli.py` is responsible for parsing the command line and config file for `CLI` input, and is the only current caller of `command.print_*()` methods. Parsing responsibility should remain in the cli module.

jmartin-tech commented on garak/command.py:
```diff
@@ -147,16 +152,46 @@ def end_run():
 
-def print_plugins(prefix: str, color):
-    from colorama import Style
+def print_plugins(prefix: str, color, filtered_plugins=None):
```
```suggestion
def print_plugins(prefix: str, color, selected_plugins=None):
```

jmartin-tech reviewed (changes requested):
The current revision still attempts to evaluate `cli` arguments in a location that should not be responsible for accessing them. See comments for further detail.

JosephDavisC commented:
I just pushed the changes and moved the CLI parsing into cli.py and updated command.py as suggested. Thanks so much for all the feedback and patience throughout this process. This is actually my first open-source PR, so I really appreciate the guidance and the time you’ve taken to help me improve it! 🙏

jmartin-tech commented on garak/command.py:
```diff
@@ -147,16 +152,46 @@ def end_run():
+    rows = enumerate_plugins(
+        category=prefix
+    )  # [("probes.dan.AntiDAN", active_bool), ...]
+    short = [(p.replace(f"{prefix}.", ""), a) for p, a in rows]
+
+    # Apply filtering if specified
+    if selected_plugins is not None:
+        # Convert selected_plugins to the same format (remove prefix)
+        filtered_short_names = [
+            p.replace(f"{prefix}.", "")
+            for p in selected_plugins
+            if p.startswith(f"{prefix}.")
+        ]
+        short = [(n, a) for (n, a) in short if n in filtered_short_names]
+
+        if not short:
+            print(f"No {prefix} match the provided filter")
+            sys.exit(2)
+
+    # Add family headers for any families that have members
+    module_names = {n.split(".")[0] for n, _ in short}
+    short += [
+        (m, True) for m in module_names
+    ]  # family lines (always active for banner)
```
I see the note here, however I am not seeing the value in having a `family` entry when passed a `selected_plugins` set. I suspect this is due to the comment from @leondz however this not the desired function.

The original issue noted the goal here is to get the list of plugin classes that will be used a runtime, a `family` is not `used` it can be provided in the spec to denote use of all `active` classes in that family.

The examples in the issue were intentional though maybe a bit unclear to other reviewers.

jmartin-tech replied:
This can be much simpler, I understand the idea to introduce guard code and the previous mention of adding _family_ however for in internal utility function this seems unnecessary.

Note this suggestion does rely on a requirement that the list of `selected_plugins` is a valid set for the plugin type requested, so something like:
``` Suggestion
    if selected_plugins is not None:
        if len(selected_plugins) > 0 and prefix in selected_plugins[0]:
            rows = zip(selected_plugins, [True] * len(selected_plugins))
        else:
            print(f"No {prefix} match the provided filter")
            return
    short = [(p.replace(f"{prefix}.", ""), a) for p, a in rows]
```

There might be a few other edge case where the message here is not 100% accurate however I don't think we need to solve for those to get this moved forward.

jmartin-tech reviewed (changes requested):
Testing shows this is really close to the desired function.  I noticed there was some confusion about the desired output.  I believe the comment original mentioning `family` inclusion was based on ensuring the changes did not impact the `default` display where all values that can validly be offered in a spec are returned with `emoji` entires to designate `family` or `inactive` plugins.

JosephDavisC commented:
Hi @jmartin-tech, I saw you made some changes to the code. Is there anything you need me to update or fix now, or should I just wait for review/merge?

jmartin-tech commented on garak/cli.py:
```diff
@@ -162,7 +162,7 @@ def main(arguments=None) -> None:
         type=str,
         default=_config.plugins.probe_spec,
-        help="list of probe names to use, or 'all' for all (default).",
+        help="list of probe names to use, or 'all' for all (default). For filtering with --list_probes, use e.g. '--list_probes -p dan' to show only 'dan' family probes.",
```
Please revert this line, the combination is only in effect if `--list_probes` or `--list_detectors` is used, the comment about combination should only be on those options.

jmartin-tech commented on garak/cli.py:
```diff
@@ -176,7 +176,7 @@ def main(arguments=None) -> None:
         type=str,
         default=_config.plugins.detector_spec,
-        help="list of detectors to use, or 'all' for all. Default is to use the probe's suggestion.",
+        help="list of detectors to use, or 'all' for all. Default is to use the probe's suggestion. For filtering with --list_detectors, use e.g. '--list_detectors -d misleading.Invalid'.",
```
Please revert this line, the combination is only in effect if `--list_probes` or `--list_detectors` is used, the comment about combination should only be on those options.

jmartin-tech commented on garak/cli.py:
```diff
@@ -224,10 +224,14 @@ def main(arguments=None) -> None:
-        "--list_probes", action="store_true", help="list available vulnerability probes"
+        "--list_probes",
+        action="store_true",
+        help="list available vulnerability probes. Usage: combine with --probes/-p to filter, e.g. '--list_probes -p dan' to show only 'dan' family probes.",
```
```suggestion
        help="list available all probes. Usage: combine with --probes/-p to filter for probes that will be activated based on a `probe_spec`, e.g. '--list_probes -p dan' to show only active 'dan' family probes.",
```

jmartin-tech commented on garak/cli.py:
```diff
@@ -224,10 +224,14 @@ def main(arguments=None) -> None:
-        "--list_detectors", action="store_true", help="list available detectors"
+        "--list_detectors",
+        action="store_true",
+        help="list available detectors. Usage: combine with --detectors/-d to filter, e.g. '--list_detectors -d misleading.Invalid' to show only that detector.",
```
```suggestion
        help="list available detectors. Usage: combine with --detectors/-d to filter for detectors that will be activated based on a `detector_spec`, e.g. '--list_detectors -d misleading.Invalid' to show only that detector.",
```

jmartin-tech reviewed (changes requested):
The help comments really only make sense when a `--list` options is passed, some ideas offered that I think are more clear about the impacts of usage and to focus the examples on when they are expected to have an impact on output.

JosephDavisC commented:
Reverted the help text change for --probes as suggested. Thanks for the feedback! Let me know if anything else needs fixing
